File: src/main/java/no/nav/pensjon/selvbetjeningopptjening/security/time/LeewayExpirationChecker.java
package no.nav.pensjon.selvbetjeningopptjening.security.time;

import no.nav.pensjon.selvbetjeningopptjening.time.TimeProvider;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Component;

import java.time.LocalDateTime;

@Component
public class LeewayExpirationChecker implements ExpirationChecker {

    private final TimeProvider timeProvider;
    private final long leewaySeconds;

    public LeewayExpirationChecker(TimeProvider timeProvider,
                                   @Value("${sts.token.expiration.leeway}") String leewaySeconds) {
        this.timeProvider = timeProvider;
        this.leewaySeconds = Long.parseLong(leewaySeconds);
    }

    public boolean isExpired(LocalDateTime issuedTime, Long expiresInSeconds) {
        LocalDateTime deadline = issuedTime.plusSeconds(expiresInSeconds - leewaySeconds);
        return timeProvider.time().isAfter(deadline);
    }

    public LocalDateTime time() {
        return timeProvider.time();
    }
}

File: src/main/java/no/nav/pensjon/selvbetjeningopptjening/tech/security/egress/token/validation/LeewayExpirationChecker.kt
package no.nav.pensjon.selvbetjeningopptjening.tech.security.egress.token.validation

import org.springframework.beans.factory.annotation.Value
import org.springframework.stereotype.Component
import java.time.LocalDateTime

/**
 * Checks expiration, allowing a certain leeway period.
 * If current time is after the start of the leeway period, the token shall be renewed,
 * even if it is not yet expired (this is to avoid 'false' expiry due to clock skew).
 */
@Component
class LeewayExpirationChecker(
    val timeProvider: TimeProvider,
    @Value("\${token.expiration.leeway}") leewaySeconds: String
) : ExpirationChecker {

    private val leeway: Long = leewaySeconds.toLong()

    override fun isExpired(issuedTime: LocalDateTime, expiresInSeconds: Long): Boolean {
        val deadline = issuedTime.plusSeconds(expiresInSeconds - leeway)
        return timeProvider.time().isAfter(deadline)
    }

    override fun time(): LocalDateTime {
        return timeProvider.time()
    }
}
